music_assistant/server/providers/snapcast/__init__.py

```diff
@@ -0,0 +1,305 @@
+                    stream.ffmpeg.terminate()
+                    self.logger.debug("ffmpeg player stopped")
+                except FFmpegError:
+                    self.logger.debug("Fail to stop ffmpeg player")
+
+    async def cmd_pause(self, player_id: str) -> None:
+        """Send PAUSE command to given player."""
+        await self.cmd_stop(player_id)
+
+    async def cmd_volume_mute(self, player_id, muted):
+        """Send MUTE command to given player."""
+        self.mass.create_task(self._snapserver.client(player_id).set_muted(muted))
+
+    async def _remove_stream(self, stream_id):
+        self.mass.create_task(self._server.stream_remove_stream(stream_id))
+
+    def _snapclient_get_group_clients_identifiers(self, player_id):
+        group = self._get_snapgroup(player_id)
+        return [ele for ele in group.clients if ele != player_id]
+
+    async def cmd_sync(self, player_id: str, target_player: str) -> None:
+        """Sync Snapcast player."""
+        child_player = self.mass.players.get(player_id)
+        assert child_player  # guard
+        parent_player = self.mass.players.get(target_player)
+        assert parent_player  # guard
+        # always make sure that the parent player is part of the sync group
+        # parent_player.group_childs.add(parent_player.player_id)
+        parent_player.group_childs.add(child_player.player_id)
+        child_player.synced_to = parent_player.player_id
+
+        group = self._get_snapgroup(target_player)
+        self.mass.create_task(group.add_client(player_id))
+
+        self.mass.players.update(child_player.player_id)
+        self.mass.players.update(parent_player.player_id)
+
+    async def cmd_unsync(self, player_id: str) -> None:
+        """Unsync Snapcast player."""
+        group = self._get_snapgroup(player_id)
+        await group.remove_client(player_id)
+        group = self._get_snapgroup(player_id)
+        stream_id = await self._get_empty_stream(player_id)
+        await group.set_stream(stream_id)
+        self._handle_update()
+
+    def _get_snapgroup(self, player_id):
+        client = self._snapserver.client(player_id)
+        return client.group
+
+    def _get_snapstream(self, player_id):
+        group = self._get_snapgroup(player_id)
+        return self._snapserver.stream(group.stream)
+
+    def _synced_to(self, player_id):
+        ret = None
+        snap_group = self._get_snapgroup(player_id)
+        snap_clients = list(filter(lambda x: x != player_id, snap_group.clients))
+        if player_id == snap_group.clients[0]:  # Player is a Sync group master
+            player = self.mass.players.get(player_id)
+            player.group_childs.clear()
+            for snap_client in snap_clients:
+                player.group_childs.add(snap_client)
+        elif len(snap_clients) > 0:
+            ret = snap_clients[0]  # Return sync group master id
+        return ret
+
+    async def _get_empty_stream(self, player_id) -> str:
```

Comment:
what does this do ? maybe add a comment why this is here, why its needed ?

Reply:
This looks for a stream in snapcast to send the audio to in case they are all used, it creates a new one, the default snapcast configuration is a named pipe.
This method tells snapcast to create a tcp socket as a server to send the audio to.  
Port 4953 is not chosen randomly. It is the default when creating a tcp socket in snapcast. 

[https://github.com/badaix/snapcast/blob/develop/doc/configuration.md#tcp-server](url)

Reply:
If you consider that it should go in the library it is not a problem, I will release a new version to simplify the imports. I could also add it to that release 
I can add parameters like codec, samplerate, bit depth, channels. 
I would look for one that is free and meets the criteria or create a new one. 
The codec is for distribution from the snapcast server to the clients. There is a client for esp32 without PSRAM that only supports Opus and pcm at the moment.

Reply:
OK, all clear! Let's go from here and improve/iterate later.